lidia/lidia.go

```diff
@@ -0,0 +1,261 @@
+// Package lidia implements a custom binary format for efficient symbolization of Go profiles.
+//
+// Lidia provides a compact binary representation of symbol information extracted from
+// ELF files, optimized for fast lookup by memory address. This is particularly useful
+// for symbolizing profile data collected from Go applications.
+package lidia
+
+import (
+	"debug/elf"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+)
+
+type ReaderAtCloser interface {
+	io.ReadCloser
+	io.ReaderAt
+}
+
+// Table represents a lidia symbol table that can be queried for lookups.
+type Table struct {
+	file ReaderAtCloser
+	hdr  header
+	opt  options
+
+	vaTable []byte
+
+	fieldsBuffer []byte
+}
+
+// SourceInfoFrame represents a single frame of symbolized profiling information.
+// It contains the name of the function, the source file path, and the line number
+// at which the profiling sample was taken.
+type SourceInfoFrame struct {
+	LineNumber   uint64
+	FunctionName string
+	FilePath     string
+}
+
+// GoRange represents a function range to be added to a lidia file.
+type GoRange struct {
+	VA        uint64
+	Length    uint32
+	Function  string
+	File      string
+	CallFile  string
+	CallLine  uint32
+	Depth     uint32
+	LineTable LineTable
+}
+
+// LineTable represents source line number information.
+type LineTable []LineTableEntry
+
+// LineTableEntry maps an offset to a line number.
+type LineTableEntry struct {
+	Offset     uint32
+	LineNumber uint32
+}
+
+// OpenReader creates a new Table from the provided ReaderAtCloser.
+// It reads the header, validates the format, and prepares the table for lookups.
+// The caller is responsible for closing the returned Table.
+func OpenReader(f ReaderAtCloser, opt ...Option) (*Table, error) {
+	var err error
+	res := new(Table)
+
+	for _, o := range opt {
+		o(&res.opt)
+	}
+
+	res.file = f
+
+	hdr, err := readHeader(f)
+	if err != nil {
+		res.Close()
+		return nil, fmt.Errorf("failed to read header: %w", err)
+	}
+
+	if hdr.magic != magic {
+		res.Close()
+		return nil, fmt.Errorf("invalid magic number: expected 0x%x, got 0x%x", magic, hdr.magic)
+	}
+
+	if hdr.version != version {
+		res.Close()
+		return nil, fmt.Errorf("unsupported version: expected %d, got %d", version, hdr.version)
+	}
+
```

Comment:
Actually, it would be nice if we allowed for forward compatibility – let's discuss it separately